### ISSUE TITLE
Electron 10,contextIsolation

### ISSUE THREAD

**Issue by Coding-Kiwi** [electron-userland/spectron#693]
I wanted to try out spectron but it seems that this is impossible if you have relatively modern standards which are:
- electron 10 (cant launch, have to use electron 9)
- nodeIntegration false (

**Comment by FrancescoBorzi** [comment]
Confirmed. It's failing for me with:

>  Error: Failed to create session.
session not created: This version of ChromeDriver only supports Chrome version 83

**Comment by Coding-Kiwi** [comment]
For me it just opens tons of electron and cmd windows, I have to abort it to make it stop, no error messages

**Comment by FrancescoBorzi** [comment]
> For me it just opens tons of electron and cmd windows, I have to abort it to make it stop, no error messages

For me also opens tons of electron windows, and in the terminal I got that error

**Comment by AMDZEN3** [comment]
Just chiming in that I'm having the same issue with the hundreds of windows opening.  After upgrading to 10 I had to set `enableRemoteModule: true` in my `webPreferences`.  In case that's important.

**Comment by FrancescoBorzi** [comment]
@jkleinsc can you please have a look at this? we are still unable to use Electron 10 because of this :(

**Comment by jkleinsc** [comment]
Spectron unfortunately is written to use the remote module and setting `contextIsolation: true` and `enableRemoteModule: false` will make spectron unusable.  As @AMDZEN3 mentioned you will need to set `enableRemoteModule: true` in the  `webPreferences` setting in order to use Spectron.  

Long term Spectron needs to be rewritten to not use the remote module.

**Comment by baparham** [comment]
@jkleinsc this workaround doesn't work to get Spectron running with electron 10 as there are a few other blockers, like the chromium webdriver version mismatch. Even with enableRemoteModule: True, and manually modifying the dependency chain of electron to use the proper chrome webdriver version, spectron still opens hundreds of windows and fails with errors.

Is there a plan to create a version of spectron (e.g. 12) that officially supports electron 10, at least until the remote module is completely deprecated?

**Comment by Coding-Kiwi** [comment]
As mentioned in the issue description I think we need something that communicates between main and renderer through a preload contextbridge. The solution to this cant be "yeah just disable every suggested security feature and you can test your code"

I switched so using puppeteer instead.

**Comment by Nawv** [comment]
@jkleinsc I'm also seeing this issue with Electron 10 opening multiple windows before failing, with the same error message about the ChromeDriver mismatch. This is with `contextIsolation: false` & `enableRemoteModule: true`.

**Comment by jkleinsc** [comment]
@Nawv can you try the newly released v12.0.0 version?

**Comment by baparham** [comment]
Looks good so far, thanks John!

**Comment by Nawv** [comment]
@jkleinsc Yes, that's working for me - cheers!

**Comment by baparham** [comment]
@Coding-Kiwi this issue is still not really resolved yet though, right? I believe that spectron still fails when `contextIsolation: true` is set.

**Comment by Coding-Kiwi** [comment]
@baparham Yes, I created this issue because we need a longterm solution. I really appreciate the 12.0.0 release but in the context of the issue, it is a make-shift solution just to get things running kinda. I understand that making spectron work without the remote module might require a major rework but I'd like this issue open so people who are experiencing issues with contextIsolation find this.

**Comment by FrancescoBorzi** [comment]
> @baparham Yes, I created this issue because we need a longterm solution. I really appreciate the 12.0.0 release but in the context of the issue, it is a make-shift solution just to get things running kinda. I understand that making spectron work without the remote module might require a major rework but I'd like this issue open so people who are experiencing issues with contextIsolation find this.

I agree with @Coding-Kiwi about the original issue being not solved yet. However, it's nice to have a workaround allowing to use Electron 10. So thanks @jkleinsc

**Comment by NoneOfMaster** [comment]
> Spectron unfortunately is written to use the remote module and setting `contextIsolation: true` and `enableRemoteModule: false` will make spectron unusable.

^ Since a developer following the latest Electron security practices during app development (and/or just having default settings) then attempting to use Spectron _will_ run into this issue and need to find this thread for the solution, I wonder if it is something that should be addressed in the README.

My approach after following the README's [Node Integration](https://github.com/electron-userland/spectron#node-integration) section was to reuse NODE_ENV to set web preferences:

```js
// main.js

const TESTING = process.env.NODE_ENV === 'test';

const envBasedWebPreferences = {
  enableRemoteModule: TESTING,
  contextIsolation: !TESTING,
};

// any new windows
new BrowserWindow({
  ...,
  webPreferences: {
    ...envBasedWebPreferences,
    ...,
  }
})
```

**Comment by JulienBier** [comment]
I tried all the solutions exposed here (including @NoneOfMaster suggestion) to use Spectron with `contextIsolation: true` but none of them work.
Here is my setup:
- electron 11.1.0
- spectron 13.0.0
- browerWindow settings:
```
webPreferences: {
	nodeIntegration: false, // tried with true
	contextIsolation: true,
	enableRemoteModule: false // tried with true
},
```
In Spectron, `window.require` is undefined.

Does anyone managed to run Spectron when contextIsolation is set to true?

**Comment by baparham** [comment]
@JulienBier I was under the impression that @NoneOfMaster suggestion was to set `contextIsolation: false` for spectron testing purposes and default back to true for his non-testing environment.

**Comment by cawa-93** [comment]
> 
> 
> > Spectron unfortunately is written to use the remote module and setting `contextIsolation: true` and `enableRemoteModule: false` will make spectron unusable.
> 
> ^ Since a developer following the latest Electron security practices during app development (and/or just having default settings) then attempting to use Spectron _will_ run into this issue and need to find this thread for the solution, I wonder if it is something that should be addressed in the README.
> 
> My approach after following the README's [Node Integration](https://github.com/electron-userland/spectron#node-integration) section was to reuse NODE_ENV to set web preferences:
> 
> ```js
> // main.js
> 
> const TESTING = process.env.NODE_ENV === 'test';
> 
> const envBasedWebPreferences = {
>   enableRemoteModule: TESTING,
>   contextIsolation: !TESTING,
> };
> 
> // any new windows
> new BrowserWindow({
>   ...,
>   webPreferences: {
>     ...envBasedWebPreferences,
>     ...,
>   }
> })
> ```

This will not work for several reasons:
- Most likely, your application in production state uses a preload script that run `electron.contextBridge.exposeInMainWorld( ... )`
- contextBridge API can only be used when `contextIsolation` is enabled.
- With `contextBridge: false` You will get an error in the rendering and the tests will fail.
- With `contextBridge: true` Tests can't work. Even if you limit yourself to the possibilities, and use only those functions that can work with `contextBridge: true`: such an elementary thing as `app.stop()` does not work and the automated test pipeline just hangs waiting for the process to complete, which cannot be completed.

**Comment by NoneOfMaster** [comment]
> * Most likely, your application in production state uses a preload script that run `electron.contextBridge.exposeInMainWorld( ... )`
> * contextBridge API can only be used when `contextIsolation` is enabled.
> * With `contextBridge: false` You will get an error in the rendering and the tests will fail.
> * With `contextBridge: true` Tests can't work.

This is right (and really helpful). I ran into it after writing a few more tests. So far, a conditional preload as follows, along with the [`main.js` above](https://github.com/electron-userland/spectron/issues/693#issuecomment-744141364) seems to be working:
```js
// preload.js
const { contextBridge, ipcRenderer } = require('electron');

const api = {
  // github.com/electron/electron/issues/21437#issuecomment-573522360
  ...safeApiMethods,
};

if (process.env.NODE_ENV === 'test') {
  window.electronRequire = require; // github.com/electron-userland/spectron#node-integration
  window.api = api;
} else {
  contextBridge.exposeInMainWorld('api', api);
}
```

**Comment by JulienBier** [comment]
Thanks for the help. I managed to get it to work by declaring the api in windows when running the e2e.
This is not ideal because the tested code is slightly different in production but it is better than nothing.

**Comment by cawa-93** [comment]
> 
> 
> > * Most likely, your application in production state uses a preload script that run `electron.contextBridge.exposeInMainWorld( ... )`
> > * contextBridge API can only be used when `contextIsolation` is enabled.
> > * With `contextBridge: false` You will get an error in the rendering and the tests will fail.
> > * With `contextBridge: true` Tests can't work.
> 
> This is right (and really helpful). I ran into it after writing a few more tests. So far, a conditional preload as follows, along with the [`main.js` above](https://github.com/electron-userland/spectron/issues/693#issuecomment-744141364) seems to be working:
> 
> ```js
> // preload.js
> const { contextBridge, ipcRenderer } = require('electron');
> 
> const api = {
>   // github.com/electron/electron/issues/21437#issuecomment-573522360
>   ...safeApiMethods,
> };
> 
> if (process.env.NODE_ENV === 'test') {
>   window.electronRequire = require; // github.com/electron-userland/spectron#node-integration
>   window.api = api;
> } else {
>   contextBridge.exposeInMainWorld('api', api);
> }
> ```

This is a good idea. However, one thing needs to be added.

`exposeInMainWorld` freezes any data before transmitting it to the main world. So to keep your identity in test mode, you need to do the same.

**Summarizing all the above, we have the following solution** (my example uses TypeScript):
```js
// main.ts

webPreferences: {
  preload:  'path/to/preload.ts',
  contextIsolation: !isTestMode,  // Spectron tests can't work with contextIsolation: true
  enableRemoteModule: isTestMode, // Spectron tests can't work with enableRemoteModule: false
},
```
```js
// preload.ts

/**
 * @see https://github.com/electron/electron/issues/21437#issuecomment-573522360
 */
const api = {
  doThing: () => ipcRenderer.send('do-a-thing'),
  ... safeApiMethods
} as const

if (isTestMode) {
  /**
   * The "Main World" is the JavaScript context that your main renderer code runs in.
   * By default, the page you load in your renderer executes code in this world.
   *
   * @see https://www.electronjs.org/docs/api/context-bridge
   */
  contextBridge.exposeInMainWorld('electron', api)
} else {
  /**
   * Recursively Object.freeze() on objects and functions
   * @see https://github.com/substack/deep-freeze
   * @param o Object on which to lock the attributes
   */
  function deepFreeze<T extends Record<string, any>>(o: T): Readonly<T> {
    Object.freeze(o)

    Object.getOwnPropertyNames(o).forEach(prop => {
      if (o.hasOwnProperty(prop)
        && o[prop] !== null
        && (typeof o[prop] === 'object' || typeof o[prop] === 'function')
        && !Object.isFrozen(o[prop])) {
        deepFreeze(o[prop])
      }
    })

    return o
  }

  deepFreeze(api)

  // @ts-expect-error https://github.com/electron-userland/spectron#node-integration
  window.electronRequire = require

  // @ts-expect-error https://github.com/electron-userland/spectron/issues/693#issuecomment-747872160
  window.electron = api
}
```

**Comment by lacymorrow** [comment]
### Edit: This no longer works in Electron versions above 11

I couldn't read `process.env.NODE_ENV` from the `preload.js` but I found I could access a property through `contextBridge.internalContextBridge.contextIsolationEnabled`:

```js
// Spectron issue: https://github.com/electron-userland/spectron/issues/693
if ( contextBridge.internalContextBridge && contextBridge.internalContextBridge.contextIsolationEnabled ) {

	/**
	   * The "Main World" is the JavaScript context that your main renderer code runs in.
	   * By default, the page you load in your renderer executes code in this world.
	   *
	   * @see https://www.electronjs.org/docs/api/context-bridge
	   */
	contextBridge.exposeInMainWorld( 'electron', api )

} else {
	// ...
	
	// deepFreeze from https://github.com/electron-userland/spectron/issues/693#issuecomment-748482545
	window.electron = deepFreeze( api )
	window.testing = true
	// Github.com/electron-userland/spectron#node-integration
	window.electronRequire = require

}
```

**Comment by cawa-93** [comment]
@lacymorrow ```contextBridge.internalContextBridge.contextIsolationEnabled``` looks like a good solution. However, I do not understand why this thing is not documented and is not in the electron type definitions? Is it safe to use this in production?

**Comment by lacymorrow** [comment]
@cawa-93 I have no idea why it's not documented, maybe is supposed to be internal but it looks like it exactly reflects the value passed to `window.webPreferences`: https://github.com/electron/electron/blob/7672aa95258bafce9abb2e38b1cf4a4d20ff32d0/lib/renderer/api/context-bridge.ts#L4

**Comment by AHDesigns** [comment]
as a small note, for anyone using webpack to transpile their typescript I had to alter `require` to stop webpack trying to convert it to it's own `_webpack_require` function (which was causing me issues.).

```ts
window.electronRequire = require
```

became -->

```ts
window.electronRequire = eval('require');
```

Very good chance it's my poor webpack config knowledge 😅 and there's probably a better fix, but hope it helps someone.

**Comment by gtritchie** [comment]
https://www.electronjs.org/docs/tutorial/using-selenium-and-webdriver#setting-up-spectron says: "Spectron is the officially supported ChromeDriver testing framework for Electron."

It seems like there should be a prominent warning that Spectron has significant issues with recent versions of Electron (i.e. this issue). Not sure what process is for requesting a docs change (where to open an issue).

**Comment by zhex900** [comment]
> > > * Most likely, your application in production state uses a preload script that run `electron.contextBridge.exposeInMainWorld( ... )`
> > > * contextBridge API can only be used when `contextIsolation` is enabled.
> > > * With `contextBridge: false` You will get an error in the rendering and the tests will fail.
> > > * With `contextBridge: true` Tests can't work.
> > 
> > 
> > This is right (and really helpful). I ran into it after writing a few more tests. So far, a conditional preload as follows, along with the [`main.js` above](https://github.com/electron-userland/spectron/issues/693#issuecomment-744141364) seems to be working:
> > ```js
> > // preload.js
> > const { contextBridge, ipcRenderer } = require('electron');
> > 
> > const api = {
> >   // github.com/electron/electron/issues/21437#issuecomment-573522360
> >   ...safeApiMethods,
> > };
> > 
> > if (process.env.NODE_ENV === 'test') {
> >   window.electronRequire = require; // github.com/electron-userland/spectron#node-integration
> >   window.api = api;
> > } else {
> >   contextBridge.exposeInMainWorld('api', api);
> > }
> > ```
> 
> This is a good idea. However, one thing needs to be added.
> 
> `exposeInMainWorld` freezes any data before transmitting it to the main world. So to keep your identity in test mode, you need to do the same.
> 
> **Summarizing all the above, we have the following solution** (my example uses TypeScript):
> 
> ```js
> // main.ts
> 
> webPreferences: {
>   preload:  'path/to/preload.ts',
>   contextIsolation: !isTestMode,  // Spectron tests can't work with contextIsolation: true
>   enableRemoteModule: isTestMode, // Spectron tests can't work with enableRemoteModule: false
> },
> ```
> 
> ```js
> // preload.ts
> 
> /**
>  * @see https://github.com/electron/electron/issues/21437#issuecomment-573522360
>  */
> const api = {
>   doThing: () => ipcRenderer.send('do-a-thing'),
>   ... safeApiMethods
> } as const
> 
> if (!isTestMode) {
>   /**
>    * The "Main World" is the JavaScript context that your main renderer code runs in.
>    * By default, the page you load in your renderer executes code in this world.
>    *
>    * @see https://www.electronjs.org/docs/api/context-bridge
>    */
>   contextBridge.exposeInMainWorld('electron', api)
> } else {
>   /**
>    * Recursively Object.freeze() on objects and functions
>    * @see https://github.com/substack/deep-freeze
>    * @param o Object on which to lock the attributes
>    */
>   function deepFreeze<T extends Record<string, any>>(o: T): Readonly<T> {
>     Object.freeze(o)
> 
>     Object.getOwnPropertyNames(o).forEach(prop => {
>       if (o.hasOwnProperty(prop)
>         && o[prop] !== null
>         && (typeof o[prop] === 'object' || typeof o[prop] === 'function')
>         && !Object.isFrozen(o[prop])) {
>         deepFreeze(o[prop])
>       }
>     })
> 
>     return o
>   }
> 
>   deepFreeze(api)
> 
>   // @ts-expect-error https://github.com/electron-userland/spectron#node-integration
>   window.electronRequire = require
> 
>   // @ts-expect-error https://github.com/electron-userland/spectron/issues/693#issuecomment-747872160
>   window.electron = api
> }
> ```
@cawa-93 Why do you need to freeze the API `deepFreeze(api)` ?

**Comment by cawa-93** [comment]
@zhex900 From [`contextBridge.exposeInMainWorld`](https://www.electronjs.org/docs/api/context-bridge#api) docs:

> Function values are proxied to the other context and all other values are **copied** and **frozen**. Any data / primitives sent in the API become immutable and updates on either side of the bridge do not result in an update on the other side.

So, if we simulate `contextBridge.exposeInMainWorld` behaviour with `window.electron = api` then, I think, We should also freeze all data.

**Comment by johns10** [comment]
Thank ya'll so much. I've literally been on this ALL DAY. I'm updating the readme with this info.

**Comment by oscartbeaumont** [comment]
I have started working on a project called [Spectroscope](https://github.com/oscartbeaumont/spectroscope) as an alternative to Spectron that doesn't require disabling all the security features and jumping through so many hoops to run your Electron tests while maintaining a secure application. Spectroscope is not compatible with the Spectron API but I'm sure if someone wanted to fix Spectron they could refer to Spectroscope for inspiration on how to.

**Comment by SNDST00M** [comment]
Could this be implemented in the electron/chromedriver repository?
